java-checks/src/main/java/org/sonar/java/checks/DateTimeFormatterMismatchCheck.java

```diff
@@ -0,0 +1,216 @@
+      if (initializer == null || !initializer.is(Tree.Kind.STRING_LITERAL)) {
+        return;
+      }
+      String pattern = ((LiteralTree) initializer).value();
+      if (isInfringingPattern(pattern)) {
+        reportIssue(invocation, CHANGE_YEAR_FORMAT_WEEK_BASED_MESSAGE);
+      }
+    }
+  }
+
+  private static boolean isInfringingPattern(String pattern) {
+    return WEEK_PATTERN.matcher(pattern).matches() && YEAR_OF_ERA_PATTERN.matcher(pattern).matches();
+  }
+
+  public static boolean refersToWeek(ExpressionTree argument) {
+    return isChronoFieldWeek(argument) || isWeekOfWeekBasedYearUsed(argument);
+  }
+
+  private static boolean isWeekOfWeekBasedYearUsed(ExpressionTree argument) {
+    if (argument.is(Tree.Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree call = (MethodInvocationTree) argument;
+      return WEEK_OF_WEEK_BASED_YEAR_MATCHER.matches(call);
+    }
+    return false;
+  }
+
+  public static boolean isChronoFieldWeek(ExpressionTree argument) {
+    if (argument.is(Tree.Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree select = (MemberSelectExpressionTree) argument;
+      IdentifierTree identifier = select.identifier();
+      return identifier.name().equals("ALIGNED_WEEK_OF_YEAR");
+    }
+    return false;
+  }
+
+  public static boolean refersToYear(ExpressionTree argument) {
+    return isChronoFieldYear(argument) || isWeekBasedYearUsed(argument);
+  }
+
+  private static boolean isWeekBasedYearUsed(ExpressionTree argument) {
+    if (argument.is(Tree.Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree call = (MethodInvocationTree) argument;
+      return WEEK_BASED_YEAR_MATCHER.matches(call);
+    }
+    return false;
+  }
+
+  public static boolean isChronoFieldYear(ExpressionTree argument) {
+    if (argument.is(Tree.Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree select = (MemberSelectExpressionTree) argument;
+      IdentifierTree identifier = select.identifier();
+      return identifier.name().equals("YEAR");
+    }
+    return false;
+  }
+
+  static class ChainVisitor extends BaseTreeVisitor {
+    private boolean usesWeek = false;
+    private boolean usesWeekOfWeekBasedYear = false;
+    private boolean usesYear = false;
+    private boolean usesWeekBasedYear = false;
+    private final List<JavaFileScannerContext.Location> locations = new ArrayList<>();
+    private ExpressionTree primary = null;
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree callToAppendValue) {
+      inspectCall(callToAppendValue);
+      ExpressionTree wanderer = callToAppendValue.methodSelect();
+      while (wanderer.is(Tree.Kind.MEMBER_SELECT)) {
+        ExpressionTree expression = ((MemberSelectExpressionTree) wanderer).expression();
+        if (!expression.is(Tree.Kind.METHOD_INVOCATION)) {
+          break;
+        }
+        MethodInvocationTree mit = (MethodInvocationTree) expression;
+        inspectCall(mit);
+        wanderer = mit.methodSelect();
```

Comment:
Can we reverse the condition on this `if` here and not use a `break` as this is not necessary anymore

Reply:
The test can only be moved higher in the while loop condition because the code we will either need to exit the loop or try to access method select on an object that does not support it.